### PR TITLE
Adjust Monitoring status detection logic

### PR DESCRIPTION
**Pull request by thxCode**
**Problem:**
After enabling Monitoring, Rancher dashboard cannot perceive cluster level Prometheus is working or not. 

**Solution:**
- Don't reconcile Conditions of `MonitoringStatus`, which cause so many useless logs, https://github.com/rancher/rancher/issues/17733. Focusing on `MonitoringEnabled` condition is enought, it needs to change UI logic, @loganhz : 

image

- Setup a handler to watch the cluster level Prometheus StatefulSet is ready or not

**Issue:**
- https://github.com/rancher/rancher/issues/17733
- https://github.com/rancher/rancher/issues/19278

### CODE REVIEW

File: pkg/controllers/user/monitoring/clusterHandler.go [hunk 1]
```diff
@@ -368,118 +365,6 @@ func (ch *clusterHandler) deployApp(appName, appTargetNamespace string, appProje
...
-
-	checkers := make([]func() error, 0, len(monitoringStatus.Conditions))
-	if !ConditionGrafanaDeployed.IsTrue(monitoringStatus) {
-		checkers = append(checkers, func() error {
-			return isGrafanaDeployed(ch.app.agentDeploymentClient, appTargetNamespace, appName, monitoringStatus, cluster.Name)
-		})
-	}
-	if enabledExporterNode && !ConditionNodeExporterDeployed.IsTrue(monitoringStatus) {
-		checkers = append(checkers, func() error {
-			return isNodeExporterDeployed(ch.app.agentDaemonSetClient, appTargetNamespace, appName, monitoringStatus)
```
Comment (orangedeng) [inline pkg/controllers/user/monitoring/clusterHandler.go]: I think here is the only place we use `isNodeExporterDeployed` function. Maybe we don't need it anymore.

Comment (orangedeng) [reply pkg/controllers/user/monitoring/clusterHandler.go]: Same as `isNodeExporterWithdrew`, `isGrafanaWithdrew`, `isKubeStateExporterWithdrew`, `isMetricExpressionDeployed`, `isKubeStateExporterDeployed` and `isGrafanaDeployed `.

File: pkg/controllers/user/monitoring/cluster_monitoring_enabled_handler.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+	cattleClustersClient mgmtv3.ClusterInterface
+}
+
+func (cmeh *clusterMonitoringEnabledHandler) sync(key string, statefulSet *v1beta2.StatefulSet) (runtime.Object, error) {
+	prometheusStatefulSetName := fmt.Sprintf("prometheus-%s", cmeh.appName)
+	if statefulSet == nil || statefulSet.DeletionTimestamp != nil || statefulSet.Name != prometheusStatefulSetName {
+		return statefulSet, nil
+	}
+
+	cluster, err := cmeh.cattleClustersClient.Get(cmeh.clusterName, metav1.GetOptions{})
```
[the remark this answers is not in the document]
Comment (thxCode) [reply pkg/controllers/user/monitoring/cluster_monitoring_enabled_handler.go]: Updated

File: pkg/controllers/user/monitoring/cluster_monitoring_enabled_handler.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+	// workload is ready, going to change condition
+	if !mgmtv3.ClusterConditionMonitoringEnabled.IsTrue(cluster) {
+		// detect Endpoints is ready or not
+		svcName, svcNamespace, _ := monitoring.ClusterPrometheusEndpoint()
+		ep, err := cmeh.appEndpointsLister.Get(svcNamespace, svcName)
+		if err != nil {
+			return statefulSet, err
+		}
+		if len(ep.Subsets) == 0 || len(ep.Subsets[0].Addresses) == 0 {
+			return statefulSet, fmt.Errorf("%s/%s endpoints is not ready", svcNamespace, svcName)
```
[the remark this answers is not in the document]
Comment (thxCode) [reply pkg/controllers/user/monitoring/cluster_monitoring_enabled_handler.go]: Updated

File: pkg/controllers/user/monitoring/cluster_monitoring_enabled_handler.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+		mgmtv3.ClusterConditionMonitoringEnabled.Message(newCluster, "")
+		mgmtv3.ClusterConditionMonitoringEnabled.Reason(newCluster, "")
+
+		return statefulSet, cmeh.updateCluster(newCluster)
+	}
+
+	return statefulSet, nil
+}
+
+func (cmeh *clusterMonitoringEnabledHandler) updateCluster(cluster *mgmtv3.Cluster) error {
```
[the remark this answers is not in the document]
Comment (thxCode) [reply pkg/controllers/user/monitoring/cluster_monitoring_enabled_handler.go]: Updated

File: pkg/controllers/user/monitoring/clusterHandler.go (hunk 1 above)
Comment (thxCode) [reply pkg/controllers/user/monitoring/clusterHandler.go]: Updated

**Comment by orangedeng** [comment]
LGTM

**Comment by loganhz** [comment]
LGTM

**Comment by loganhz** [comment]
We need to merge the UI PR at the same time.

https://github.com/rancher/rancher/issues/19197

File: pkg/controllers/user/monitoring/clusterHandler.go [hunk 5]
```diff
@@ -99,34 +100,47 @@ func (ch *clusterHandler) doSync(cluster *mgmtv3.Cluster) error {
...
-		appAnswers, err := ch.deployApp(appName, appTargetNamespace, appProjectName, cluster, etcdTLSConfigs, systemComponentMap)
+		_, err = ch.deployApp(appName, appTargetNamespace, appProjectName, cluster, etcdTLSConfigs, systemComponentMap)
 		if err != nil {
 			mgmtv3.ClusterConditionMonitoringEnabled.Unknown(cluster)
 			mgmtv3.ClusterConditionMonitoringEnabled.Message(cluster, err.Error())
 			return errors.Wrap(err, "failed to deploy monitoring")
 		}
 
-		if err := ch.detectAppComponentsWhileInstall(appName, appTargetNamespace, cluster, appAnswers); err != nil {
+		if err := ch.deployMetrics(cluster); err != nil {
```
Comment (orangedeng) [inline pkg/controllers/user/monitoring/clusterHandler.go]: Is there any possible not to deploy it every sync?

File: pkg/controllers/user/monitoring/clusterHandler.go [hunk 6]
```diff
@@ -368,134 +382,83 @@ func (ch *clusterHandler) deployApp(appName, appTargetNamespace string, appProje
...
+func (ch *clusterHandler) isPrometheusReady(cluster *mgmtv3.Cluster) error {
+	svcName, namespace, _ := monitoring.ClusterPrometheusEndpoint()
+	endpoints, err := ch.agentEndpointsLister.Get(namespace, svcName)
+	if err != nil {
+		return errors.Wrapf(err, "prometheus is not ready")
+	}
+
+	if len(endpoints.Subsets) == 0 || len(endpoints.Subsets[0].Addresses) == 0 {
+		// return ForgetError as clusterMonitoringEnabledHandler will trigger a new loop
+		return &controller.ForgetError{Err: fmt.Errorf("prometheus is not ready")}
```
Comment (orangedeng) [inline pkg/controllers/user/monitoring/clusterHandler.go]: Please don't mix `get an error` and `prometheus nor ready` up as error.

File: pkg/controllers/user/monitoring/clusterHandler.go (hunk 5 above)
Comment (thxCode) [reply pkg/controllers/user/monitoring/clusterHandler.go]: Updated

File: pkg/controllers/user/monitoring/clusterHandler.go (hunk 6 above)
Comment (orangedeng) [reply pkg/controllers/user/monitoring/clusterHandler.go]: Please change to 
```
func (ch *clusterHandler) isPrometheusReady(cluster *mgmtv3.Cluster) (bool,error) 
```

Comment (orangedeng) [reply pkg/controllers/user/monitoring/clusterHandler.go]: It's not necessary to use `controller.ForgetError` here.

Comment (thxCode) [reply pkg/controllers/user/monitoring/clusterHandler.go]: Updated

**Comment by orangedeng** [comment]
LGTM

**Pull request merged**